File: leetcode_problems/typescript/src/lc108_sorted_arr_to_bst.ts
class TreeNode1 {
  val: number;
  left: TreeNode1 | null;
  right: TreeNode1 | null;
  constructor(val?: number, left?: TreeNode1 | null, right?: TreeNode1 | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function inorder_travesal(
  node: TreeNode1 | null,
  hook: CallableFunction
): void {
  if (node === null) {
    return;
  }

  inorder_travesal(node.left, hook);
  hook(node);
  inorder_travesal(node.right, hook);
}

function sortedArrayToBST(
  nums: number[],
  start: number | null = null,
  end: number | null = null
): TreeNode1 | null {
  if (start === null) {
    start = 0;
  }
  if (end === null) {
    end = nums.length - 1;
  }
  if (start > end) {
    return null;
  }

  let mid = start + Math.floor((end - start) / 2);
  let node = new TreeNode1(nums[mid]);

  node.left = sortedArrayToBST(nums, start, mid - 1);
  node.right = sortedArrayToBST(nums, mid + 1, end);

  return node;
}

function printNode(node: TreeNode1): void {
  if (node === null) {
    return;
  }
  console.log(node.val);
}

File: leetcode_problems/cpp/lc108_sorted_arr_to_bst.cpp
#include <vector>
using std::vector;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *build(vector<int> &nums, int start, int end)
    {
        if (start > end)
        {
            return nullptr;
        }

        int mid = start + (end - start) / 2;
        TreeNode *node = new TreeNode(nums[mid]);
        node->left = build(nums, start, mid - 1);
        node->right = build(nums, mid + 1, end);
        return node;
    }

    TreeNode *sortedArrayToBST(vector<int> &nums)
    {
        int n = nums.size();
        return build(nums, 0, n - 1);
    }
};
